File: re-implementation/tests/testthat/test-initialization.R
# Tests for MAGICAL_initialization.

context("MAGICAL_initialization")

test_that("returns a list with the documented 11 named fields", {
  fx <- make_synthetic_fixture(seed = 11L)
  init <- fx$Initial_model

  expected <- c("T_prior", "T_mean", "T_var",
                "B_prior", "B_mean", "B_var", "B_prob",
                "L_prior", "L_mean", "L_var", "L_prob")
  expect_named(init, expected, ignore.order = TRUE)
})

test_that("T_prior equals input TF_log2Count", {
  fx <- make_synthetic_fixture(seed = 12L)
  expect_equal(fx$Initial_model$T_prior, fx$Candidate_circuits$TF_log2Count)
})

test_that("T_mean equals input TF_log2Count", {
  fx <- make_synthetic_fixture(seed = 13L)
  expect_equal(fx$Initial_model$T_mean, fx$Candidate_circuits$TF_log2Count)
})

test_that("T_var has shape (M, S) and finite non-negative values", {
  fx <- make_synthetic_fixture(seed = 14L)
  M <- nrow(fx$Candidate_circuits$TFs)
  S <- length(fx$loaded_data$Common_samples)
  expect_equal(dim(fx$Initial_model$T_var), c(M, S))
  expect_true(all(is.finite(fx$Initial_model$T_var)))
  expect_true(all(fx$Initial_model$T_var >= 0))
})

test_that("B_prior has shape (P, M) and zero where the candidate prior is zero", {
  fx <- make_synthetic_fixture(seed = 15L)
  P <- nrow(fx$Candidate_circuits$Peaks)
  M <- nrow(fx$Candidate_circuits$TFs)
  expect_equal(dim(fx$Initial_model$B_prior), c(P, M))

  cand <- as.matrix(fx$Candidate_circuits$TF_Peak_Binding)
  expect_true(all(fx$Initial_model$B_prior[cand == 0] == 0))
})

test_that("B_prob has values in [0, 1] and is zero outside the candidate mask", {
  fx <- make_synthetic_fixture(seed = 16L)
  bp <- fx$Initial_model$B_prob
  expect_true(all(is.finite(bp)))
  expect_true(all(bp >= 0 & bp <= 1))

  cand <- as.matrix(fx$Candidate_circuits$TF_Peak_Binding)
  expect_true(all(bp[cand == 0] == 0))
})

test_that("B_var has shape (1, M) and finite positive values", {
  fx <- make_synthetic_fixture(seed = 17L)
  M <- nrow(fx$Candidate_circuits$TFs)
  expect_equal(dim(fx$Initial_model$B_var), c(1L, M))
  expect_true(all(is.finite(fx$Initial_model$B_var)))
  expect_true(all(fx$Initial_model$B_var > 0))
})

test_that("L_prior has shape (P, G) and zero where the candidate prior is zero", {
  fx <- make_synthetic_fixture(seed = 18L)
  P <- nrow(fx$Candidate_circuits$Peaks)
  G <- nrow(fx$Candidate_circuits$Genes)
  expect_equal(dim(fx$Initial_model$L_prior), c(P, G))

  cand <- as.matrix(fx$Candidate_circuits$Peak_Gene_looping)
  expect_true(all(fx$Initial_model$L_prior[cand == 0] == 0))
})

test_that("L_prob has values in [0, 1] and is zero outside the candidate mask", {
  fx <- make_synthetic_fixture(seed = 19L)
  lp <- fx$Initial_model$L_prob
  expect_true(all(is.finite(lp)))
  expect_true(all(lp >= 0 & lp <= 1))

  cand <- as.matrix(fx$Candidate_circuits$Peak_Gene_looping)
  expect_true(all(lp[cand == 0] == 0))
})

test_that("L_var is a finite non-negative scalar", {
  fx <- make_synthetic_fixture(seed = 20L)
  lv <- fx$Initial_model$L_var
  expect_true(length(lv) == 1L)
  expect_true(is.finite(lv))
  expect_true(lv >= 0)
})

test_that("initialization is deterministic under the same seed", {
  # MAGICAL_initialization has no RNG, so it should be a pure function of inputs.
  fx1 <- make_synthetic_fixture(seed = 21L)
  fx2 <- make_synthetic_fixture(seed = 21L)
  expect_equal(fx1$Initial_model, fx2$Initial_model)
})

File: tests/test_initialization.py
"""Tests for ``magical.initialization`` mirroring the R suite."""

from __future__ import annotations

import numpy as np


def test_returns_initial_model_with_all_expected_fields(fixture) -> None:
    im = fixture.initial_model
    for name in (
        "T_prior",
        "T_mean",
        "T_var",
        "B_prior",
        "B_mean",
        "B_var",
        "B_prob",
        "L_prior",
        "L_mean",
        "L_var",
        "L_prob",
    ):
        assert hasattr(im, name), f"missing field {name}"


def test_T_prior_and_T_mean_match_input_tf_log2count(fixture) -> None:
    tf_log2 = fixture.candidate_circuits.TF_log2Count
    np.testing.assert_allclose(fixture.initial_model.T_prior, tf_log2)
    np.testing.assert_allclose(fixture.initial_model.T_mean, tf_log2)


def test_T_var_is_broadcast_across_samples(fixture) -> None:
    im = fixture.initial_model
    T_var = im.T_var
    # Every column of T_var should equal the row-variance of TF_log2Count
    for s in range(T_var.shape[1]):
        np.testing.assert_allclose(T_var[:, 0], T_var[:, s])


def test_B_prior_zero_outside_candidate_binding(fixture) -> None:
    im = fixture.initial_model
    mask = np.asarray(fixture.candidate_circuits.TF_Peak_Binding.todense()) > 0
    assert np.all(im.B_prior[~mask] == 0.0)


def test_B_prob_is_in_zero_one_at_candidate_positions(fixture) -> None:
    im = fixture.initial_model
    mask = np.asarray(fixture.candidate_circuits.TF_Peak_Binding.todense()) > 0
    vals = im.B_prob[mask]
    # B_prob = 1 - p-value, so must lie in [0, 1]
    assert vals.min() >= 0.0
    assert vals.max() <= 1.0


def test_B_var_is_row_shape_and_positive(fixture) -> None:
    im = fixture.initial_model
    M = fixture.candidate_circuits.TFs.shape[0]
    assert im.B_var.shape == (1, M)
    assert np.all(im.B_var > 0)


def test_L_prior_zero_outside_candidate_looping(fixture) -> None:
    im = fixture.initial_model
    mask = np.asarray(fixture.candidate_circuits.Peak_Gene_looping.todense()) > 0
    assert np.all(im.L_prior[~mask] == 0.0)


def test_L_prob_is_in_zero_one_at_candidate_positions(fixture) -> None:
    im = fixture.initial_model
    mask = np.asarray(fixture.candidate_circuits.Peak_Gene_looping.todense()) > 0
    vals = im.L_prob[mask]
    assert vals.min() >= 0.0
    assert vals.max() <= 1.0


def test_L_var_is_positive_scalar(fixture) -> None:
    L_var = fixture.initial_model.L_var
    assert isinstance(L_var, float)
    assert L_var > 0


def test_dimensions_match_candidate_shapes(fixture) -> None:
    im = fixture.initial_model
    cc = fixture.candidate_circuits
    M = cc.TFs.shape[0]
    P = cc.Peaks.shape[0]
    G = cc.Genes.shape[0]
    S = len(fixture.loaded_data.common_samples)
    assert im.T_prior.shape == (M, S)
    assert im.T_var.shape == (M, S)
    assert im.B_prior.shape == (P, M)
    assert im.B_prob.shape == (P, M)
    assert im.L_prior.shape == (P, G)
    assert im.L_prob.shape == (P, G)


def test_initialization_is_deterministic(make_fixture) -> None:
    a = make_fixture(seed=42)
    b = make_fixture(seed=42)
    np.testing.assert_allclose(a.initial_model.B_prior, b.initial_model.B_prior)
    np.testing.assert_allclose(a.initial_model.L_prior, b.initial_model.L_prior)
    np.testing.assert_allclose(a.initial_model.T_var, b.initial_model.T_var)
